### ISSUE TITLE
Refresh TOTP code automatically if it has short time to live

### ISSUE THREAD

**Issue by ksteckert** [Yubico/yubioath-desktop#210]
Haven't looked through the code yet to see how it's engineered, so this might not be possible if the device itself generates the TOTP code upon touch, rather than providing the secret to the app.

Sometimes when I generate a TOTP, it expires within a matter of seconds, which isn't enough time to paste it on the site and submit, which means I have to double-click it again, and touch the Yubikey again.  This isn't a big deal... but it would be great if the Authenticator app would check the remaining time upon touch/generation, and if the validation is going to expire within a certain threshold (maybe within 4 or 5 seconds) then it would automatically update with the next iteration code (like the apps on the phone do) without having to click & touch again.

Basically, if it's obvious we won't have enough time to input the generated code, refresh it automatically (only once, not indefinitely) instead of making us double-click and touch again.

If the Yubikey itself releases the TOTP code upon touch, then this probably isn't something you can do, by design, and I'm perfectly fine knowing how secure it is-- but if it's actually releasing the shared secret when you touch the yubikey, well then you have everything you need to generate the next code, and I would appreciate some serious consideration of this request.

You could make this an optional feature in settings, that the users could toggle on if they want it.

**Comment by emlun** [comment]
Thanks for he suggestion! The YubiKey does not release the shared secret, so it's not possible to "save the touch" and regenerate another code after the expiry period refreshes. What we _could_ probably do is delay sending the command to the YubiKey at all until the period refreshes. I'll take a look at how much work that would be, and we'll decide then if it's something we want to do.

**Comment by ksteckert** [comment]
Glad to know the shared secret is secure on the Yubikey, and I appreciate the consideration.  I definitely wouldn't mind waiting a few seconds before touching the key if I knew it was waiting for a fresh TOTP code, but I could see some users thinking it was malfunctioning because a prompt to press the button didn't show up when they clicked on it.

Does the yubikey know the timestamp natively thru a driver interface or something, or does the Authenticator app send that with the request?  My thought is if the time right now was 18:42:27, when you send the request to the yubikey, could you send the NEXT timestamp (18:42:30 instead of 18:42:00)?  I can see this as an awkward way to implement it (generating a future code), so it'd probably be a feature a user would explicity enable in the settings.  To denote if the code is a future code or not, maybe display it in a different color (yellow, orange, light gray??) so the user knows not to actually submit it until it turns the normal BLACK.  When the green bar gets to the end, the text color could change to the normal BLACK, and the green bar would start again.

I attached an example of what the setting could look like, if this is feasible and something you want to implement.

![yubico_settings_next-code](https://user-images.githubusercontent.com/28814650/33047549-6324259a-ce24-11e7-9c62-192e4dd7904e.png)

**Comment by ksteckert** [comment]
Yet another idea to help alleviate frustration, involves the green bar at the top that shows how much time is left.  You could have an option in settings to ALWAYS show the bar, but when there are no active codes, the bar changes to gray instead of green.  This way there would be a visual indicator, so I can clearly see that I should wait a couple seconds before generating the code.  This might actually be the easiest way to combat the frustration.

It probably wouldn't involve much more than adding a boolean for the setting, and checking it in 2 or 3 places. 

When the code to 'hide' the progress (upon completion) bar gets called, it would check the setting, and if enabled, change the color of the bar to grey and start it again, otherwise it would operate as usual.

The code that normally starts the bar after TOTP generation would probably just need to add a command to switch the color back to green.

**Comment by dainnilsson** [comment]
The way this is done on the Android app, which is probably also how it should be done here, is that when a TOTP code is triggered at time T, the app checks the validity period. If this is less than 10 seconds, it will automatically generate a code for the next period instead of the current one (in practice it just always uses T+10).

In theory this means that the codes shown might not be valid for up to 10 seconds, but in practice with the time it takes to copy the code, plus the server typically allowing for some clock drift, this isn't a problem. This also means that the timer bar sometimes animates over a slightly longer period than 30 seconds.

As the YubiKey is persistently connected in the desktop case, this would only apply to touch-enabled TOTP credentials.

**Comment by emlun** [comment]
Implemented as described by @dainnilsson above!
